### ISSUE TITLE
Support multiple layers for managing imports

### ISSUE THREAD

**Issue by mikelmaron** [openstreetmap/iD#2586]
JOSM allows multiple data layers, which is handy when doing a merge of import data. Makes it easy to compare vs existing OSM data, and you can copy/paste features or tags between layers. For iD, simply allowing to have an OSM layer, and a "reference" layer (perhaps just a GeoJSON file passed by URL or from the local computer), could suffice.

**Comment by bhousel** [comment]
As an intermediate step towards this, I think it makes sense to give the GPX layer a better name (suggestions welcome) and support a few other file formats. I'm kind of surprised iD doesn't support geojson here. 

Then, actually giving the user some ability to import those features as OSM objects would be cool too, if we can keep it simple and do it in a way that doesn't break anything.

**Comment by mikelmaron** [comment]
@bhousel makes sense. first step sounds pretty straightforward? and second step, what would it take to make the "(suggestions welcome)" layer interactive, clickable, and copyable?

**Comment by bhousel** [comment]
> what would it take to make the "(suggestions welcome)" layer interactive, clickable, and copyable?

I don't think that we would need to make the reference layer editable.. We could just add a button to import whatever is there (gpx/kml/geojson) onto the main map where it will be editable.  Some potential issues:

* Imported features need to have presets assigned
* Imported features often need simplification and we don't have an operation for that (yet)
* Ensure that the user understands not to import improperly licensed data
* Consensus that this is a feature that should be built in the first place.  OSM community tends to be pretty anti-import lately.

But anyway I like the idea.  I have a bunch of gpx tracks sitting on my computer that I am too lazy to trace.

**Comment by pnorman** [comment]
Is this in scope for a basic editor? I recall imports an external data sources being out of I'd scope in the past.

**Comment by ajturner** [comment]
I think that this would be very helpful - at least from the concept story: 

**"As a User, when editing in iD, I can add layers from other Services that I can edit and push into OpenStreetMap"**

- iD should keep it's core data model focused on OSM nodes/ways/relations.
- a plugin would be loaded at runtime that adds support for API conversion
- a URL to a GeoJSON file would be handled by this plugin
- plugin parses the GeoJSON and converts it into OSM nodes/ways and hands it to iD core - output is same as if the user had hand-drawn the objects
- User can now interact with the objects and commit to OSM

This is something I'd like to help out on.

**Comment by systemed** [comment]
This has long been a feature of P2 - you can load various file formats (.shp, .gpx, .kml, .osm) which appear as "background layers", and are invisibly converted into ways and nodes on import. As background layers they don't do anything, but alt-clicking a feature in one of them (in traditional Potlatch obscure-keypress fashion) brings it through to the main editable layer as a live feature.

**Comment by slibby** [comment]
We have opened a new issue to discuss this topic for adding web-service based vector data from ArcGIS GeoServices (ArcGIS Server Map Services and Feature Services) as background + editable layers: https://github.com/openstreetmap/iD/issues/4164. I would like to think that the new issue gets at some of these requests and is similar to the function of uploading a shapefile or geojson document to iD and using that as a base layer. Adding shapefile and GeoJSON support would be a straightforward addition to the work we have done so far if it was of interest.

**Comment by bhousel** [comment]
Awesome @slibby - yes I am also seeing #4164 as a great starting point towards improving how imports are done and would supersede this issue.

I have to comment on that thread, but was holding off to let everyone have their say and also because I've been focused on getting 2.3.2 released. 👍

**Comment by tyrasd** [comment]
> "reference" layer (perhaps just a GeoJSON file)

this is already possible since a long time now.

> Is this in scope for a basic editor?

I would say probably not. Proper _imports_ are probably better done with a editor like JOSM with proper support for it.
